File: python/app/repository.py
import hashlib
import json
import secrets
from typing import Any
from datetime import datetime, timedelta

from sqlalchemy import func, or_, select
from sqlalchemy.dialects.mysql import insert
from sqlalchemy.exc import IntegrityError
from sqlalchemy.orm import Session, selectinload

from .models import AdminUser, CallbackEvent, Subtask, Task


TERMINAL_STATUSES = {"completed", "partial_completed", "failed", "stopped"}


class TaskRepository:
    def __init__(self, db: Session):
        self.db = db

    def save_submitted_task(self, request: dict[str, Any], response: dict[str, Any]) -> None:
        task_id = response["taskId"]
        self._upsert_task(
            {
                "task_id": task_id,
                "status": response.get("status", "pending"),
                "prompts_json": request.get("prompts") or [],
                "platforms_json": request.get("platforms") or [],
                "region_code_json": request.get("regionCode") or [],
                "callback_url": response.get("callbackUrl"),
                "total_items": response.get("totalTask") or response.get("totalItems") or 0,
                "completed_items": 0,
                "failed_items": 0,
                "poll_url": response.get("pollUrl"),
                "raw_request_json": request,
                "raw_response_json": response,
            }
        )
        self._upsert_subtasks(task_id, response.get("subTaskList") or [])
        self.db.commit()

    def save_remote_result(self, payload: dict[str, Any]) -> None:
        task_id = payload["taskId"]
        self._upsert_task(
            {
                "task_id": task_id,
                "status": payload.get("status", "processing"),
                "prompts_json": [],
                "platforms_json": [],
                "region_code_json": [],
                "callback_url": payload.get("callbackUrl"),
                "total_items": payload.get("totalItems") or 0,
                "completed_items": payload.get("completedItems") or 0,
                "failed_items": payload.get("failedItems") or 0,
                "poll_url": payload.get("pollUrl"),
                "created_at": payload.get("createdAt"),
                "completed_at": payload.get("completedAt"),
                "raw_request_json": {},
                "raw_response_json": payload,
            },
            update_fields={
                "status",
                "total_items",
                "completed_items",
                "failed_items",
                "callback_url",
                "poll_url",
                "created_at",
                "completed_at",
                "raw_response_json",
                "updated_at",
            },
        )
        self._upsert_subtasks(task_id, payload.get("subTaskList") or [])
        self._backfill_task_metadata(task_id)
        self.db.commit()

    def save_callback(self, payload: dict[str, Any]) -> bool:
        raw = json.dumps(payload, ensure_ascii=False, sort_keys=True)
        event = CallbackEvent(
            task_id=payload["taskId"],
            payload_json=payload,
            payload_hash=hashlib.sha256(raw.encode("utf-8")).hexdigest(),
            process_status="processed",
            processed_at=func.now(),
        )
        self.db.add(event)
        try:
            self.db.flush()
        except IntegrityError:
            self.db.rollback()
            return True
        self.save_remote_result(payload)
        return False

    def list_tasks(self, page: int, size: int, status: str | None) -> dict[str, Any]:
        query = select(Task)
        count_query = select(func.count()).select_from(Task)
        if status:
            query = query.where(Task.status == status)
            count_query = count_query.where(Task.status == status)
        items = self.db.scalars(query.order_by(Task.created_local_at.desc()).limit(size).offset((page - 1) * size)).all()
        total = self.db.scalar(count_query)
        return {"items": [self._task_to_dict(item, False) for item in items], "total": total, "page": page, "size": size}

    def get_task(self, task_id: str) -> dict[str, Any] | None:
        task = self.db.scalar(select(Task).options(selectinload(Task.subtasks)).where(Task.task_id == task_id))
        if not task:
            return None
        data = self._task_to_dict(task, True)
        events = self.db.scalars(select(CallbackEvent).where(CallbackEvent.task_id == task_id).order_by(CallbackEvent.received_at.desc()).limit(20)).all()
        data["callbackEvents"] = [self._callback_event_to_dict(event) for event in events]
        return data

    def unfinished_task_ids(self, limit: int = 20) -> list[str]:
        rows = self.db.execute(
            select(Task.task_id)
            .outerjoin(Subtask, Subtask.task_id == Task.task_id)
            .where(
                or_(
                    Task.status.not_in(TERMINAL_STATUSES),
                    Subtask.subtask_id.is_(None),
                    Subtask.status.is_(None),
                    Subtask.status.not_in(TERMINAL_STATUSES),
                    Task.status.in_({"completed", "partial_completed"})
                    & (Subtask.status == "completed")
                    & ((Subtask.answer_content.is_(None)) | (Subtask.answer_content == "")),
                )
            )
            .group_by(Task.task_id)
            .order_by(func.min(Task.created_local_at).asc())
            .limit(max(limit, 1))
        ).all()
        return [row[0] for row in rows]

    def find_admin_by_username(self, username: str) -> AdminUser | None:
        return self.db.scalar(select(AdminUser).where(AdminUser.username == username.strip(), AdminUser.status == 1))

    def issue_token(self, user: AdminUser, ip: str | None) -> tuple[str, datetime]:
        token = secrets.token_hex(32)
        expires_at = datetime.now() + timedelta(days=7)
        user.auth_token_hash = hashlib.sha256(token.encode("utf-8")).hexdigest()
        user.token_expires_at = expires_at
        user.last_login_at = datetime.now()
        user.last_login_ip = ip
        self.db.commit()
        return token, expires_at

    def authenticate(self, token: str | None) -> AdminUser | None:
        if not token:
            return None
        token_hash = hashlib.sha256(token.strip().encode("utf-8")).hexdigest()
        user = self.db.scalar(select(AdminUser).where(AdminUser.auth_token_hash == token_hash, AdminUser.status == 1))
        if not user or not user.token_expires_at or user.token_expires_at < datetime.now():
            return None
        return user

    def logout(self, token: str | None) -> None:
        user = self.authenticate(token)
        if user:
            user.auth_token_hash = None
            user.token_expires_at = None
            self.db.commit()

    def _upsert_task(self, values: dict[str, Any], update_fields: set[str] | None = None) -> None:
        now = datetime.now()
        values.setdefault("created_local_at", now)
        values["updated_at"] = now
        stmt = insert(Task).values(**values)
        allowed_update_fields = update_fields or set(values.keys())
        update_values = {
            key: stmt.inserted[key]
            for key in values.keys()
            if key not in {"task_id", "created_local_at"} and key in allowed_update_fields
        }
        self.db.execute(stmt.on_duplicate_key_update(**update_values))

    def _upsert_subtasks(self, task_id: str, rows: list[dict[str, Any]]) -> None:
        for row in rows:
            subtask_id = row.get("subTaskId")
            if not subtask_id:
                continue
            values = {
                "subtask_id": subtask_id,
                "task_id": task_id,
                "platform": row.get("platform"),
                "mode": row.get("mode"),
                "prompt": row.get("prompt"),
                "status": row.get("status"),
                "time": row.get("time"),
                "page_screenshot": row.get("pageScreenshot"),
                "answer_content": row.get("answerContent"),
                "reference_list_json": row.get("referenceList") or [],
                "citation_list_json": row.get("citationList") or [],
                "reasoning_process_json": row.get("reasoningProcess") or {},
                "recommended_questions_json": row.get("recommendedQuestions") or [],
                "media_content_json": row.get("mediaContent") or [],
                "error_message": row.get("errorMessage"),
                "proxy_ip": row.get("proxyIp"),
                "raw_result_json": row,
                "updated_at": datetime.now(),
            }
            stmt = insert(Subtask).values(**values)
            update_values = {key: stmt.inserted[key] for key in values.keys() if key != "subtask_id"}
            self.db.execute(stmt.on_duplicate_key_update(**update_values))

    def _backfill_task_metadata(self, task_id: str) -> None:
        task = self.db.scalar(select(Task).where(Task.task_id == task_id))
        if not task:
            return

        subtasks = self.db.scalars(select(Subtask).where(Subtask.task_id == task_id).order_by(Subtask.updated_at.asc())).all()
        if not subtasks:
            return

        changed = False
        if not task.prompts_json:
            prompts: list[str] = []
            for subtask in subtasks:
                if subtask.prompt and subtask.prompt not in prompts:
                    prompts.append(subtask.prompt)
            if prompts:
                task.prompts_json = prompts
                changed = True

        if not task.platforms_json:
            platform_keys: set[tuple[str, str]] = set()
            platforms: list[dict[str, str]] = []
            for subtask in subtasks:
                if not subtask.platform or not subtask.mode:
                    continue
                key = (subtask.platform, subtask.mode)
                if key in platform_keys:
                    continue
                platform_keys.add(key)
                platforms.append({"platform": subtask.platform, "mode": subtask.mode})
            if platforms:
                task.platforms_json = platforms
                changed = True

        if changed:
            task.updated_at = datetime.now()

    def _task_to_dict(self, task: Task, include_subtasks: bool) -> dict[str, Any]:
        data = {column.name: getattr(task, column.name) for column in Task.__table__.columns}
        data["taskId"] = data.pop("task_id")
        data["task_id"] = data["taskId"]
        for key in ("prompts_json", "platforms_json", "region_code_json", "raw_request_json", "raw_response_json"):
            data[key] = json.dumps(data.get(key), ensure_ascii=False)
        if include_subtasks:
            data["subTaskList"] = [self._subtask_to_dict(item) for item in task.subtasks]
        return data

    def _subtask_to_dict(self, subtask: Subtask) -> dict[str, Any]:
        data = {column.name: getattr(subtask, column.name) for column in Subtask.__table__.columns}
        data["subTaskId"] = data.pop("subtask_id")
        data["subtask_id"] = data["subTaskId"]
        for key in ("reference_list_json", "citation_list_json", "reasoning_process_json", "recommended_questions_json", "media_content_json", "raw_result_json"):
            data[key] = json.dumps(data.get(key), ensure_ascii=False)
        return data

    def _callback_event_to_dict(self, event: CallbackEvent) -> dict[str, Any]:
        data = {column.name: getattr(event, column.name) for column in CallbackEvent.__table__.columns}
        data["payload_json"] = json.dumps(data.get("payload_json"), ensure_ascii=False)
        return data


def public_user(user: AdminUser) -> dict[str, Any]:
    return {"id": user.id, "username": user.username, "displayName": user.display_name, "role": user.role}

File: rust/src/repository.rs
use serde_json::{Value, json};
use sqlx::{MySqlPool, Row};

use crate::util::{
    get_datetime, get_i64, get_json_string, get_opt_datetime, get_opt_i64, get_opt_json_string,
    get_opt_string, get_string, json_text, now_string, sha256,
};

/// Lists tasks for the management UI with optional status filtering.
pub(crate) async fn list_tasks(
    pool: &MySqlPool,
    page: u64,
    size: u64,
    status: Option<String>,
) -> anyhow::Result<(i64, Vec<Value>)> {
    let offset = ((page - 1) * size) as i64;
    let limit = size as i64;
    let (total, rows) = if let Some(status) = status {
        let total: i64 = sqlx::query_scalar("SELECT COUNT(*) FROM geo_tasks WHERE status=?")
            .bind(&status)
            .fetch_one(pool)
            .await?;
        let rows = sqlx::query(
            "SELECT * FROM geo_tasks WHERE status=? ORDER BY created_local_at DESC LIMIT ? OFFSET ?",
        )
        .bind(&status)
        .bind(limit)
        .bind(offset)
        .fetch_all(pool)
        .await?;
        (total, rows)
    } else {
        let total: i64 = sqlx::query_scalar("SELECT COUNT(*) FROM geo_tasks")
            .fetch_one(pool)
            .await?;
        let rows =
            sqlx::query("SELECT * FROM geo_tasks ORDER BY created_local_at DESC LIMIT ? OFFSET ?")
                .bind(limit)
                .bind(offset)
                .fetch_all(pool)
                .await?;
        (total, rows)
    };
    Ok((total, rows.iter().map(task_row).collect()))
}

/// Loads one task with subtasks and recent callback events for the detail page.
pub(crate) async fn task_detail(pool: &MySqlPool, task_id: &str) -> anyhow::Result<Option<Value>> {
    let row = sqlx::query("SELECT * FROM geo_tasks WHERE task_id=?")
        .bind(task_id)
        .fetch_optional(pool)
        .await?;
    let Some(row) = row else {
        return Ok(None);
    };

    let mut task = task_row(&row);
    let subtasks =
        sqlx::query("SELECT * FROM geo_subtasks WHERE task_id=? ORDER BY updated_at DESC")
            .bind(task_id)
            .fetch_all(pool)
            .await?;
    let events = sqlx::query(
        "SELECT * FROM geo_callback_events WHERE task_id=? ORDER BY received_at DESC LIMIT 20",
    )
    .bind(task_id)
    .fetch_all(pool)
    .await?;

    task["subTaskList"] = Value::Array(subtasks.iter().map(subtask_row).collect());
    task["callbackEvents"] = Value::Array(events.iter().map(callback_row).collect());
    Ok(Some(task))
}

/// Persists a freshly submitted task and its initial subtask summary.
pub(crate) async fn save_submitted(
    pool: &MySqlPool,
    request: &Value,
    response: &Value,
) -> anyhow::Result<()> {
    let now = now_string();
    let task_id = response.get("taskId").and_then(Value::as_str).unwrap_or("");
    sqlx::query("INSERT INTO geo_tasks (task_id,status,prompts_json,platforms_json,region_code_json,callback_url,total_items,completed_items,failed_items,poll_url,raw_request_json,raw_response_json,created_local_at,updated_at) VALUES (?,?,?,?,?,?,?,?,?,?,?,?,?,?) ON DUPLICATE KEY UPDATE status=VALUES(status),prompts_json=VALUES(prompts_json),platforms_json=VALUES(platforms_json),region_code_json=VALUES(region_code_json),callback_url=VALUES(callback_url),total_items=VALUES(total_items),poll_url=VALUES(poll_url),raw_request_json=VALUES(raw_request_json),raw_response_json=VALUES(raw_response_json),updated_at=VALUES(updated_at)")
        .bind(task_id)
        .bind(response.get("status").and_then(Value::as_str).unwrap_or("pending"))
        .bind(json_text(request.get("prompts").unwrap_or(&json!([]))))
        .bind(json_text(request.get("platforms").unwrap_or(&json!([]))))
        .bind(json_text(request.get("regionCode").unwrap_or(&json!([]))))
        .bind(response.get("callbackUrl").and_then(Value::as_str).or_else(|| request.get("callbackUrl").and_then(Value::as_str)))
        .bind(response.get("totalTask").or_else(|| response.get("totalItems")).and_then(Value::as_i64).unwrap_or(0))
        .bind(0)
        .bind(0)
        .bind(response.get("pollUrl").and_then(Value::as_str))
        .bind(json_text(request))
        .bind(json_text(response))
        .bind(&now)
        .bind(&now)
        .execute(pool)
        .await?;
    upsert_subtasks(
        pool,
        task_id,
        response.get("subTaskList").and_then(Value::as_array),
        &now,
    )
    .await?;
    Ok(())
}

/// Persists a status/result/callback payload returned by Molizhishu.
pub(crate) async fn save_remote_result(pool: &MySqlPool, payload: &Value) -> anyhow::Result<()> {
    let now = now_string();
    let task_id = payload.get("taskId").and_then(Value::as_str).unwrap_or("");
    sqlx::query("INSERT INTO geo_tasks (task_id,status,prompts_json,platforms_json,region_code_json,total_items,completed_items,failed_items,created_at,completed_at,raw_request_json,raw_response_json,created_local_at,updated_at) VALUES (?,?,?,?,?,?,?,?,?,?,?,?,?,?) ON DUPLICATE KEY UPDATE status=VALUES(status),total_items=VALUES(total_items),completed_items=VALUES(completed_items),failed_items=VALUES(failed_items),created_at=VALUES(created_at),completed_at=VALUES(completed_at),raw_response_json=VALUES(raw_response_json),updated_at=VALUES(updated_at)")
        .bind(task_id)
        .bind(payload.get("status").and_then(Value::as_str).unwrap_or("processing"))
        .bind("[]")
        .bind("[]")
        .bind("[]")
        .bind(payload.get("totalItems").and_then(Value::as_i64).unwrap_or(0))
        .bind(payload.get("completedItems").and_then(Value::as_i64).unwrap_or(0))
        .bind(payload.get("failedItems").and_then(Value::as_i64).unwrap_or(0))
        .bind(payload.get("createdAt").and_then(Value::as_i64))
        .bind(payload.get("completedAt").or_else(|| payload.get("timestamp")).and_then(Value::as_i64))
        .bind("{}")
        .bind(json_text(payload))
        .bind(&now)
        .bind(&now)
        .execute(pool)
        .await?;
    upsert_subtasks(
        pool,
        task_id,
        payload.get("subTaskList").and_then(Value::as_array),
        &now,
    )
    .await?;
    Ok(())
}

/// Records a callback payload idempotently and applies its latest task state.
pub(crate) async fn save_callback(pool: &MySqlPool, payload: &Value) -> anyhow::Result<bool> {
    let raw = json_text(payload);
    let hash = sha256(&raw);
    let task_id = payload.get("taskId").and_then(Value::as_str).unwrap_or("");
    let exists: i64 = sqlx::query_scalar("SELECT COUNT(*) FROM geo_callback_events WHERE task_id=? AND payload_hash=? AND process_status='processed'")
        .bind(task_id).bind(&hash).fetch_one(pool).await?;
    if exists > 0 {
        sqlx::query("INSERT INTO geo_callback_events (task_id,payload_json,payload_hash,process_status,received_at,processed_at) VALUES (?,?,?,'duplicate',NOW(),NOW())")
            .bind(task_id).bind(raw).bind(hash).execute(pool).await?;
        return Ok(true);
    }
    sqlx::query("INSERT INTO geo_callback_events (task_id,payload_json,payload_hash,process_status,received_at,processed_at) VALUES (?,?,?,'processed',NOW(),NOW())")
        .bind(task_id).bind(raw).bind(hash).execute(pool).await?;
    save_remote_result(pool, payload).await?;
    Ok(false)
}

/// Returns task IDs that still need status/result compensation.
pub(crate) async fn unfinished_task_ids(
    pool: &MySqlPool,
    limit: i64,
) -> anyhow::Result<Vec<String>> {
    let rows = sqlx::query_scalar::<_, String>("SELECT t.task_id FROM geo_tasks t LEFT JOIN geo_subtasks s ON s.task_id = t.task_id WHERE t.status NOT IN ('completed', 'partial_completed', 'failed', 'stopped') OR s.subtask_id IS NULL OR s.status IS NULL OR s.status NOT IN ('completed', 'partial_completed', 'failed', 'stopped') OR (t.status IN ('completed', 'partial_completed') AND s.status = 'completed' AND (s.answer_content IS NULL OR s.answer_content = '')) GROUP BY t.task_id ORDER BY MIN(t.created_local_at) ASC LIMIT ?")
        .bind(limit.max(1))
        .fetch_all(pool)
        .await?;
    Ok(rows)
}

async fn upsert_subtasks(
    pool: &MySqlPool,
    task_id: &str,
    rows: Option<&Vec<Value>>,
    now: &str,
) -> anyhow::Result<()> {
    for row in rows.into_iter().flatten() {
        let subtask_id = row.get("subTaskId").and_then(Value::as_str).unwrap_or("");
        if subtask_id.is_empty() {
            continue;
        }
        sqlx::query("INSERT INTO geo_subtasks (subtask_id,task_id,platform,mode,prompt,status,time,page_screenshot,answer_content,reference_list_json,citation_list_json,reasoning_process_json,recommended_questions_json,media_content_json,error_message,proxy_ip,raw_result_json,updated_at) VALUES (?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?,?) ON DUPLICATE KEY UPDATE platform=VALUES(platform),mode=VALUES(mode),prompt=VALUES(prompt),status=VALUES(status),time=VALUES(time),page_screenshot=VALUES(page_screenshot),answer_content=VALUES(answer_content),reference_list_json=VALUES(reference_list_json),citation_list_json=VALUES(citation_list_json),reasoning_process_json=VALUES(reasoning_process_json),recommended_questions_json=VALUES(recommended_questions_json),media_content_json=VALUES(media_content_json),error_message=VALUES(error_message),proxy_ip=VALUES(proxy_ip),raw_result_json=VALUES(raw_result_json),updated_at=VALUES(updated_at)")
            .bind(subtask_id)
            .bind(task_id)
            .bind(row.get("platform").and_then(Value::as_str))
            .bind(row.get("mode").and_then(Value::as_str))
            .bind(row.get("prompt").and_then(Value::as_str))
            .bind(row.get("status").and_then(Value::as_str))
            .bind(row.get("time").and_then(Value::as_i64))
            .bind(row.get("pageScreenshot").and_then(Value::as_str))
            .bind(row.get("answerContent").and_then(Value::as_str))
            .bind(json_text(row.get("referenceList").unwrap_or(&json!([]))))
            .bind(json_text(row.get("citationList").unwrap_or(&json!([]))))
            .bind(json_text(row.get("reasoningProcess").unwrap_or(&Value::Null)))
            .bind(json_text(row.get("recommendedQuestions").unwrap_or(&json!([]))))
            .bind(json_text(row.get("mediaContent").unwrap_or(&json!([]))))
            .bind(row.get("errorMessage").and_then(Value::as_str))
            .bind(row.get("proxyIp").and_then(Value::as_str))
            .bind(json_text(row))
            .bind(now)
            .execute(pool)
            .await?;
    }
    Ok(())
}

/// Converts a task database row to the frontend-compatible JSON shape.
pub(crate) fn task_row(row: &sqlx::mysql::MySqlRow) -> Value {
    json!({
        "task_id": get_string(row, "task_id"), "status": get_string(row, "status"),
        "prompts_json": get_json_string(row, "prompts_json"), "platforms_json": get_json_string(row, "platforms_json"),
        "region_code_json": get_json_string(row, "region_code_json"), "callback_url": get_opt_string(row, "callback_url"),
        "total_items": get_i64(row, "total_items"), "completed_items": get_i64(row, "completed_items"),
        "failed_items": get_i64(row, "failed_items"), "poll_url": get_opt_string(row, "poll_url"),
        "created_at": get_opt_i64(row, "created_at"), "completed_at": get_opt_i64(row, "completed_at"),
        "raw_request_json": get_json_string(row, "raw_request_json"), "raw_response_json": get_opt_json_string(row, "raw_response_json"),
        "last_error": get_opt_string(row, "last_error"), "created_local_at": get_datetime(row, "created_local_at"),
        "updated_at": get_datetime(row, "updated_at")
    })
}

/// Converts a subtask database row to the frontend-compatible JSON shape.
pub(crate) fn subtask_row(row: &sqlx::mysql::MySqlRow) -> Value {
    json!({
        "subtask_id": get_string(row, "subtask_id"), "task_id": get_string(row, "task_id"), "platform": get_opt_string(row, "platform"),
        "mode": get_opt_string(row, "mode"), "prompt": get_opt_string(row, "prompt"), "status": get_opt_string(row, "status"),
        "time": get_opt_i64(row, "time"), "page_screenshot": get_opt_string(row, "page_screenshot"),
        "answer_content": get_opt_string(row, "answer_content"), "reference_list_json": get_opt_json_string(row, "reference_list_json"),
        "citation_list_json": get_opt_json_string(row, "citation_list_json"), "reasoning_process_json": get_opt_json_string(row, "reasoning_process_json"),
        "recommended_questions_json": get_opt_json_string(row, "recommended_questions_json"), "media_content_json": get_opt_json_string(row, "media_content_json"),
        "error_message": get_opt_string(row, "error_message"), "proxy_ip": get_opt_string(row, "proxy_ip"),
        "raw_result_json": get_opt_json_string(row, "raw_result_json"), "updated_at": get_datetime(row, "updated_at")
    })
}

/// Converts a callback event row to the frontend-compatible JSON shape.
pub(crate) fn callback_row(row: &sqlx::mysql::MySqlRow) -> Value {
    json!({
        "id": get_i64(row, "id"), "task_id": get_string(row, "task_id"), "payload_json": get_json_string(row, "payload_json"),
        "payload_hash": get_string(row, "payload_hash"), "process_status": get_string(row, "process_status"),
        "error_message": get_opt_string(row, "error_message"), "received_at": get_datetime(row, "received_at"),
        "processed_at": get_opt_datetime(row, "processed_at")
    })
}

/// Returns the public user payload used by the shared frontend login flow.
pub(crate) fn public_user(row: &sqlx::mysql::MySqlRow) -> Value {
    json!({"id": row.try_get::<u64, _>("id").unwrap_or_default(), "username": get_string(row, "username"), "displayName": get_string(row, "display_name"), "role": get_string(row, "role")})
}
